practice/rain.py

```diff
@@ -1,86 +1,214 @@
+    False
+    """
+    try:
+        int(value)
+        return True
+    except ValueError:
+        return False
+
+
+def get_most_daily_rain_inches(columns):
     """Gets the value of the most rain for any given date
-    >>> get_most_rain_inches([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12']])
-    '.06'
+
+    >>> get_most_daily_rain_inches([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12'], \
+    ['4', '7', '6', '1']])
+    '0.07'
     """
-    most_rain = max(columns[1])
+    rainy_days = [int(value) for value in columns[3] if contains_numeral(value)]
+    most_rain = max(rainy_days)
     most_rain_inches = str(int(most_rain) / 100)
     return most_rain_inches
 
 
 def get_index_of_max_daily_rain_box(columns):
     """Takes in rows of data, and returns index of day with most rain columns max rain
-    >>> get_day_with_most_rain([['1', '2', '3', '8'], ['4', '5', '6', '1'], ['7', '8', '9', '12']])
-    '2'
+
+    >>> get_index_of_max_daily_rain_box(TEST_COLUMNS)
+    0
     """
-    index_of_max_daily_rain = columns[1].index(max(columns[1]))
-    return index_of_max_daily_rain
+    rainy_days = [int(value) for value in columns[3] if contains_numeral(value)]
+    daily_rain_values = [value for value in rainy_days]
+    day_index_of_max_daily_rain = daily_rain_values.index(max(daily_rain_values))
+    return day_index_of_max_daily_rain
 
 
-def get_day_with_most_rain (rows, columns):
+def get_day_with_most_rain(rows, columns):
+    """Returns the date of day with the most rain
+
+    >>> get_day_with_most_rain(TEST_ROWS, TEST_COLUMNS)
+    'MAR 23 2016'
+    """
     index_of_day_cell = get_index_of_max_daily_rain_box(columns)
     day_full_line = rows[index_of_day_cell]
-    date = day_full_line[0]
+    date = day_full_line[1] + ' ' + day_full_line[0] + ' ' + day_full_line[2]
     return date
 
-def output(most_rain, date_with_most_rain):
-    print('The day with the most rain was ' + date_with_most_rain + ', which had ' + most_rain + ' inches of rain.')
+
+def output_for_max_rain_date(most_rain, date_with_most_rain):
+    """Prints the day with the most rain, and how much rain fell in inches.
+
+    >>> output_for_max_rain_date('5','30-MAR-2016')
+    The day with the most rain was 30-MAR-2016, which had 5 inches of rain!
+    """
+    print('The day with the most rain was {}, which had {} inches of rain!'.format(date_with_most_rain, most_rain))
+
+
+def get_year(row):
+    """Calculates Grouping Key.
+
+    >>> get_year(['23', 'Mar', '2016'])
+    '2016'
+    """
+    return row[2]
+
+
+def get_year_to_rows(rows):
+    """Creates a dictionary of rows of data keyed to each year.
+
+    >>> get_year_to_rows([['23', 'Mar', '2016'], ['21', 'Feb', '2016']])
+    {'2016': [['23', 'Mar', '2016'], ['21', 'Feb', '2016']]}
+    """
+    year_to_rows = {
+        group_key: list(grouped_rows)
+        for group_key, grouped_rows
+        in groupby(rows, get_year)
+        }
+    return year_to_rows
```

Comment:
This function is useful, but could be made much more useful if you get it to also support carrying around the data you want grouped so that you have all of the measurements bucketed by year. See my previous comment.